### PR TITLE
Debug studio

### PR DESCRIPTION
**Pull request by zolinthecow**
## Motivation

When developing apps (especially more complex ones) with SGLang's frontend it's a bit hard to tell what is actually getting passed into the `generate/` endpoint. This PR introduces a "prompt debugging studio" that allows you to turn on a debug mode and will start intercepting generate requests and saving them to a debug server, where you can go to a web interface and view all the prompts and responses for SGL functions.

For example, if you start the debug server:

```bash
python -m sglang.launch_debug_server
```

It will start a debug server on `http://0.0.0.0:56765`

Then, you can add a debug region to an sgl function:

```python
@sgl.function
def text_qa(s, question):
    s.begin_debug_region("TEXT_QA")
    s += "Q: " + question + "\n"
    s += "A:" + sgl.gen("answer", stop="\n")

state = text_qa.run(
    question="What is the capital of France?",
    temperature=0.1,
    stream=True
)
```

The `s.begin_debug_region` will start debugging under the prompt name "TEXT_QA". You can then go to `http://localhost:56765` (will have to ssh forward if you're developing on a remote server) and you should see prompts start appearing

image

## Modifications

- Adds BeginDebugRegion and EndDebugRegion to SGLExpr for the frontend language
- Adds `launch_debug_server` which under the hood relies on `enochian-studio` (one of my packages) for the debug server
  - `enochian-studio` can be found [here](https://github.com/zolinthecow/enochian-js/tree/master/studio), I can say it's up to date since I maintain it, along with Etched
- Adds posting to the debug server if enabled for all backends

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by zolinthecow** [comment]
^ Couldn't reproduce CI error locally :(

**Comment by zhaochenyang20** [comment]
@zolinthecow Nice work! I will start to review this PR this weekend. Stay tuned!

File: docs/frontend/frontend.md [hunk 1]
```diff
@@ -233,6 +233,35 @@ def chat_example(s):
     s += sgl.assistant_end()
 ```
 
+### Debug Studio
+
+The frontend also provides a debug studio to view what exactly is getting passed into the runtime endpoint's generation API.
+To use it, first start the debug server:
+
+```bash
+python -m sglang.launch_debug_server
```
Comment (merrymercy) [inline docs/frontend/frontend.md:242]:
```suggestion
python -m sglang.lang.launch_debug_server
```

File: python/sglang/README.md [hunk 2]
```diff
@@ -9,4 +9,5 @@
 - `bench_serving.py`: Benchmark online serving with dynamic requests.
 - `global_config.py`: The global configs and constants.
 - `launch_server.py`: The entry point for launching the local server.
+- `launch_debug_server.py`: The entry point for launching the debug server + web app
```
Comment (merrymercy) [inline python/sglang/README.md:12]: This is an experimental feature for frontend language only, so please move it under `python/sglang/lang`.

File: python/sglang/lang/backend/vertexai.py [hunk 3]
```diff
@@ -50,13 +52,37 @@ def generate(
                 if s.cur_images
                 else s.text_
             )
+
+        debug_request_id = str(uuid.uuid4())
+        s.log_debug(
```
Comment (merrymercy) [inline python/sglang/lang/backend/vertexai.py:57]: It is better to add `s.log_debug` to `interpreter.py` so we do not need to do it for every backend file.

**Review by merrymercy — CHANGES_REQUESTED** [review]
Overall, it looks good! I left a few minor comments.
One thing we want to ensure is that it does not introduce any overhead when debug mode is turned off (e.g., use as few function calls and if statements as possible at runtime).

File: python/sglang/lang/backend/anthropic.py [hunk 4]
```diff
@@ -67,6 +94,20 @@ def generate_stream(
...
+                {
+                    "id": debug_request_id,
+                    "requestPrompt": str(
+                        [{"role": "system", "content": system}] + messages
+                    ),
+                    "requestTimestamp": datetime.now().isoformat(),
+                    "requestMetadata": sampling_params.to_anthropic_kwargs(),
+                }
+            ]
+        )
```
Comment (merrymercy) [inline python/sglang/lang/backend/anthropic.py:109]: This is not efficient enough. When debug is turned off, you still run the code to construct the argument, which takes some time. Please minimize the overhead and do not construct any objects when debug is turned off.

**Review by merrymercy — CHANGES_REQUESTED** [review]

**Comment by merrymercy** [comment]
closed due to inactivity and low priority

**Pull request closed without merging**